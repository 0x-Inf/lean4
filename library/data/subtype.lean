-- Copyright (c) 2014 Microsoft Corporation. All rights reserved.
-- Released under Apache 2.0 license as described in the file LICENSE.
-- Author: Leonardo de Moura, Jeremy Avigad

import logic.inhabited logic.eq logic.decidable

open decidable

structure subtype {A : Type} (P : A → Prop) :=
tag :: (elt_of : A) (has_property : P elt_of)

notation `{` binders:55 `|` r:(scoped:1 P, subtype P) `}` := r

namespace subtype
  variables {A : Type} {P : A → Prop}

  theorem tag_irrelevant {a : A} (H1 H2 : P a) : tag a H1 = tag a H2 :=
  rfl

  theorem tag_eq {a1 a2 : A} {H1 : P a1} {H2 : P a2} (H3 : a1 = a2) : tag a1 H1 = tag a2 H2 :=
  eq.subst H3 (take H2, tag_irrelevant H1 H2) H2

  protected theorem equal {a1 a2 : {x | P x}} : ∀(H : elt_of a1 = elt_of a2), a1 = a2 :=
  destruct a1 (take x1 H1, destruct a2 (take x2 H2 H, tag_eq H))

  protected definition is_inhabited [instance] {a : A} (H : P a) : inhabited {x | P x} :=
  inhabited.mk (tag a H)

  protected definition has_decidable_eq [instance] (H : decidable_eq A) : decidable_eq {x | P x} :=
  take a1 a2 : {x | P x},
    have H1 : (a1 = a2) ↔ (elt_of a1 = elt_of a2), from
      iff.intro (assume H, eq.subst H rfl) (assume H, equal H),
    decidable_iff_equiv _ (iff.symm H1)
end subtype
