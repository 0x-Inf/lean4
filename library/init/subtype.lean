/-
Copyright (c) 2014 Microsoft Corporation. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Author: Leonardo de Moura, Jeremy Avigad
-/
prelude
import init.datatypes init.logic
open decidable

universe variables u

structure subtype {A : Type u} (p : A → Prop) :=
tag :: (elt_of : A) (has_property : p elt_of)

namespace subtype

  definition exists_of_subtype {A : Type u} {p : A → Prop} : { x \ p x } → ∃ x, p x
  | ⟨a, h⟩ := ⟨a, h⟩

  variables {A : Type u} {p : A → Prop}

  theorem tag_irrelevant {a : A} (h1 h2 : p a) : tag a h1 = tag a h2 :=
  rfl

  protected theorem eq : ∀ {a1 a2 : {x \ p x}}, elt_of a1 = elt_of a2 → a1 = a2
  | ⟨x, h1⟩ ⟨.x, h2⟩ rfl := rfl

end subtype

open subtype

variables {A : Type u} {p : A → Prop}

attribute [instance]
protected definition subtype.is_inhabited {a : A} (h : p a) : inhabited {x \ p x} :=
⟨⟨a, h⟩⟩
