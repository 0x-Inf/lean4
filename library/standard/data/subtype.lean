-- Copyright (c) 2014 Microsoft Corporation. All rights reserved.
-- Released under Apache 2.0 license as described in the file LICENSE.
-- Author: Leonardo de Moura, Jeremy Avigad

import logic.classes.inhabited logic.connectives.eq

inductive subtype {A : Type} (P : A → Prop) : Type :=
| tag : Πx : A, P x → subtype P

notation `{` binders `|` r:(scoped P, subtype P) `}` := r

namespace subtype

section
  parameter {A : Type}
  parameter {P : A → Prop}

  -- TODO: make this a coercion?
  definition  elt_of (a : {x | P x}) : A := subtype_rec (λ x y, x) a
  theorem has_property (a : {x | P x}) : P (elt_of a) := subtype_rec (λ x y, y) a

  theorem elt_of_tag (a : A) (H : P a) : elt_of (tag a H) = a := refl a

  theorem subtype_destruct {Q : {x | P x} → Prop} (a : {x | P x})
      (H : ∀(x : A) (H1 : P x), Q (tag x H1)) : Q a :=
    subtype_rec H a

  theorem tag_irrelevant {a : A} (H1 H2 : P a) : tag a H1 = tag a H2 := refl (tag a H1)

  theorem tag_elt_of (a : subtype P) : Π(H : P (elt_of a)), tag (elt_of a) H = a :=
    subtype_destruct a (take (x : A) (H1 : P x) (H2 : P x), refl _)

  theorem tag_eq {a1 a2 : A} {H1 : P a1} {H2 : P a2} (H3 : a1 = a2) : tag a1 H1 = tag a2 H2 :=
  (show ∀(H2 : P a2), tag a1 H1 = tag a2 H2, from subst H3 (take H2, tag_irrelevant H1 H2)) H2

  theorem subtype_eq {a1 a2 : {x | P x}} : ∀(H : elt_of a1 = elt_of a2), a1 = a2 :=
  subtype_destruct a1 (take x1 H1, subtype_destruct a2 (take x2 H2 H, tag_eq H))

  -- TODO: need inhabited

end

end subtype

-- instance subtype_inhabited
