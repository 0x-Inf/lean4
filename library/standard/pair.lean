-- Copyright (c) 2014 Microsoft Corporation. All rights reserved.
-- Released under Apache 2.0 license as described in the file LICENSE.
-- Author: Leonardo de Moura
import logic

inductive pair (A : Type) (B : Type) : Type :=
| mk_pair : A → B → pair A B

section
  parameter {A : Type}
  parameter {B : Type}

  definition fst [inline] (p : pair A B) := pair_rec (λ x y, x) p
  definition snd [inline] (p : pair A B) := pair_rec (λ x y, y) p

  theorem pair_inhabited [instance] (H1 : inhabited A) (H2 : inhabited B) : inhabited (pair A B)
  := inhabited_elim H1 (λ a, inhabited_elim H2 (λ b, inhabited_intro (mk_pair a b)))

  theorem fst_mk_pair (a : A) (b : B) : fst (mk_pair a b) = a
  := refl a

  theorem snd_mk_pair (a : A) (b : B) : snd (mk_pair a b) = b
  := refl b

  theorem pair_ext (p : pair A B) : mk_pair (fst p) (snd p) = p
  := pair_rec (λ x y, refl (mk_pair x y)) p
end

-- notation for n-ary tuples
notation `(` h `,` t:(foldl `,` (e r, mk_pair r e) h) `)` := t
