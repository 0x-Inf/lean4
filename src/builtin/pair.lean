-- Copyright (c) 2014 Microsoft Corporation. All rights reserved.
-- Released under Apache 2.0 license as described in the file LICENSE.
-- Author: Leonardo de Moura
definition pair {A : (Type U)} {B : (Type U)} (a : A) (b : B) := tuple a, b
theorem    pair_inj1 {A : (Type U)} {B : A → (Type U)} {a b : sig x, B x} (H : a = b) : proj1 a = proj1 b
:= subst (refl (proj1 a)) H
theorem    pair_inj2 {A B : (Type U)} {a b : A # B} (H : a = b) : proj2 a = proj2 b
:= subst (refl (proj2 a)) H
theorem    pairext_proj {A B : (Type U)} {p : A # B} {a : A} {b : B} (H1 : proj1 p = a) (H2 : proj2 p = b) : p = (pair a b)
:= @pairext A (λ x, B) p (pair a b) H1 (to_heq H2)
