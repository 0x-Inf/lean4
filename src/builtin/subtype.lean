-- Copyright (c) 2014 Microsoft Corporation. All rights reserved.
-- Released under Apache 2.0 license as described in the file LICENSE.
-- Author: Leonardo de Moura
import macros
-- Simulate "subtypes" using Sigma types and proof irrelevance
definition subtype (A : (Type U)) (P : A → Bool) := sig x, P x

namespace subtype
definition rep {A : (Type U)} {P : A → Bool} (a : subtype A P) : A
:= proj1 a

definition abst {A : (Type U)} {P : A → Bool} (r : A) (H : inhabited (subtype A P)) : subtype A P
:= ε H (λ a, rep a = r)

theorem subtype_inhabited {A : (Type U)} {P : A → Bool} (H : ∃ x, P x) : inhabited (subtype A P)
:= obtain (w : A) (Hw : P w), from H,
     inhabited_intro (tuple (subtype A P) : w, Hw)

theorem P_rep {A : (Type U)} {P : A → Bool} (a : subtype A P) : P (rep a)
:= proj2 a

theorem rep_inj {A : (Type U)} {P : A → Bool} {a b : subtype A P} (H : rep a = rep b) : a = b
:= pairext H (hproof_irrel (proj2 a) (proj2 b))

theorem ex_abst {A : (Type U)} {P : A → Bool} {r : A} (H : P r) : ∃ a, rep a = r
:= exists_intro (tuple (subtype A P) : r, H) (refl r)

theorem abst_rep {A : (Type U)} {P : A → Bool} (H : inhabited (subtype A P)) (a : subtype A P)
                 : abst (rep a) H = a
:= let s1 : rep (abst (rep a) H) = rep a  :=
                @eps_ax (subtype A P) H (λ x, rep x = rep a) a (refl (rep a))
   in rep_inj s1

theorem rep_abst {A : (Type U)} {P : A → Bool} (H : inhabited (subtype A P)) : ∀ r, P r → rep (abst r H) = r
:= take r, assume Hl : P r,
     @eps_ax (subtype A P) H (λ x, rep x = r) (tuple (subtype A P) : r, Hl) (refl r)

theorem abst_inj {A : (Type U)} {P : A → Bool} (H : inhabited (subtype A P)) {r r' : A} :
                 P r → P r' → abst r H = abst r' H → r = r'
:= assume Hr Hr' Heq,
      calc r    = rep (abst r  H)  :  symm (rep_abst H r Hr)
           ...  = rep (abst r' H)  :  { Heq }
           ...  = r'               :  rep_abst H r' Hr'

theorem ex_rep {A : (Type U)} {P : A → Bool} (H : inhabited (subtype A P)) :
               ∀ a, ∃ r, abst r H = a ∧ P r
:= take a, exists_intro (rep a) (and_intro (abst_rep H a) (proj2 a))

set_opaque rep     true
set_opaque abst    true
end -- namespace subtype
set_opaque subtype true
