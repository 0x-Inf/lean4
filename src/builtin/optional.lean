-- Copyright (c) 2014 Microsoft Corporation. All rights reserved.
-- Released under Apache 2.0 license as described in the file LICENSE.
-- Author: Leonardo de Moura
import macros
import subtype
using subtype
-- We are encoding the (optional A) as a subset of A → Bool where
--   none     is the predicate that is false everywhere
--   some(a)  is the predicate that is true only at a
definition optional_pred (A : (Type U)) := (λ p : A → Bool, (∀ x, ¬ p x) ∨ exists_unique p)
definition optional (A : (Type U)) := subtype (A → Bool) (optional_pred A)

namespace optional
theorem some_pred {A : (Type U)} (a : A) : optional_pred A (λ x, x = a)
:= or_intror
     (∀ x, ¬ x = a)
     (exists_intro a
       (and_intro (refl a) (take y, assume H : y ≠ a, H)))

theorem none_pred (A : (Type U)) : optional_pred A (λ x, false)
:= or_introl (take x, not_false_trivial) (exists_unique (λ x, false))

theorem inhab (A : (Type U)) : inhabited (optional A)
:= subtype_inhabited (exists_intro (λ x, false) (none_pred A))

definition none {A : (Type U)} : optional A
:= abst (λ x, false) (inhab A)

definition some {A : (Type U)} (a : A) : optional A
:= abst (λ x, x = a) (inhab A)

definition is_some {A : (Type U)} (n : optional A) := ∃ x : A, some x = n

theorem injectivity {A : (Type U)} {a a' : A} : some a = some a' → a = a'
:= assume Heq : some a = some a',
   have eq_reps : (λ x, x = a) = (λ x, x = a'),
      from abst_inj (inhab A) (some_pred a) (some_pred a') Heq,
   show a = a',
      from (congr1 a eq_reps) ◂ (refl a)

theorem distinct {A : (Type U)} (a : A) : some a ≠ none
:= not_intro (assume N : some a = none,
   have eq_reps : (λ x, x = a) = (λ x, false),
      from abst_inj (inhab A) (some_pred a) (none_pred A) N,
   show false,
      from (congr1 a eq_reps) ◂ (refl a))

definition value {A : (Type U)} (n : optional A) (H : is_some n) : A
:= ε (inhabited_ex_intro H) (λ x, some x = n)

theorem is_some_some {A : (Type U)} (a : A) : is_some (some a)
:= exists_intro a (refl (some a))

theorem not_is_some_none {A : (Type U)} : ¬ is_some (@none A)
:= not_intro (assume N : is_some (@none A),
   obtain (w : A) (Hw : some w = @none A),
      from N,
   show false,
      from absurd Hw (distinct w))

theorem value_some {A : (Type U)} (a : A) (H : is_some (some a)) : value (some a) H = a
:= have eq1  : some (value (some a) H) = some a,
      from eps_ax (inhabited_ex_intro H) a (refl (some a)),
   show value (some a) H = a,
      from injectivity eq1

theorem false_pred {A : (Type U)} {p : A → Bool} (H : ∀ x, ¬ p x) : p = λ x, false
:= funext (λ x, eqf_intro (H x))

theorem singleton_pred {A : (Type U)} {p : A → Bool} {w : A} (H : p w ∧ ∀ y, y ≠ w → ¬ p y) : p = (λ x, x = w)
:= funext (take x, iff_intro
     (assume Hpx : p x,
      show x = w,
         from refute (assume N : x ≠ w,
                      show false,
                         from absurd Hpx (and_elimr H x N)))
     (assume Heq : x = w,
      show p x,
         from subst (and_eliml H) (symm Heq)))

theorem dichotomy {A : (Type U)} (n : optional A) : n = none ∨ ∃ a, n = some a
:= have pred : optional_pred A (rep n),
      from P_rep n,
   show n = none ∨ ∃ a, n = some a,
      from or_elim pred
        (assume Hl : ∀ x : A, ¬ rep n x,
         have rep_n_eq     : rep n    = λ x, false,
            from false_pred Hl,
         have rep_none_eq  : rep none = λ x, false,
            from rep_abst (inhab A) (λ x, false) (none_pred A),
         show n = none ∨ ∃ a, n = some a,
            from or_introl (rep_inj (trans rep_n_eq (symm rep_none_eq)))
                           (∃ a, n = some a))
        (assume Hr : ∃ x, rep n x ∧ ∀ y, y ≠ x → ¬ rep n y,
         obtain (w : A) (Hw : rep n w ∧ ∀ y, y ≠ w → ¬ rep n y),
            from Hr,
         have rep_n_eq    : rep n        = λ x, x = w,
            from singleton_pred Hw,
         have rep_some_eq : rep (some w) = λ x, x = w,
            from rep_abst (inhab A) (λ x, x = w) (some_pred w),
         have n_eq_some   : n = some w,
            from rep_inj (trans rep_n_eq (symm rep_some_eq)),
         show n = none ∨ ∃ a, n = some a,
            from or_intror (n = none)
                           (exists_intro w n_eq_some))

theorem induction {A : (Type U)} {P : optional A → Bool} (H1 : P none) (H2 : ∀ x, P (some x)) : ∀ n, P n
:= take n, or_elim (dichotomy n)
    (assume Heq : n = none,
     show P n,
        from subst H1 (symm Heq))
    (assume Hex : ∃ a, n = some a,
     obtain (w : A) (Hw : n = some w),
        from Hex,
     show P n,
        from subst (H2 w) (symm Hw))

set_opaque some           true
set_opaque none           true
set_opaque is_some        true
set_opaque value          true
end

set_opaque optional       true
set_opaque optional_pred  true
definition optional_inhabited := optional::inhab
add_rewrite optional::is_some_some optional::not_is_some_none optional::distinct optional::value_some
