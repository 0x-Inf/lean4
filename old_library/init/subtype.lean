/-
Copyright (c) 2014 Microsoft Corporation. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Author: Leonardo de Moura, Jeremy Avigad
-/
prelude
import init.datatypes init.logic
open decidable

universe variables u

structure subtype {A : Type u} (P : A → Prop) :=
tag :: (elt_of : A) (has_property : P elt_of)

notation `{` binder ` \ `:0 r:(scoped:1 P, subtype P) `}` := r

namespace subtype

  definition exists_of_subtype {A : Type u} {P : A → Prop} : { x \ P x } → ∃ x, P x
  | ⟨a, h⟩ := ⟨a, h⟩

  variables {A : Type u} {P : A → Prop}

  theorem tag_irrelevant {a : A} (h1 h2 : P a) : tag a h1 = tag a h2 :=
  rfl

  protected theorem eq : ∀ {a1 a2 : {x \ P x}}, elt_of a1 = elt_of a2 → a1 = a2
  | ⟨x, h1⟩ ⟨.x, h2⟩ rfl := rfl

end subtype

open subtype

variables {A : Type u} {P : A → Prop}

attribute [instance]
protected definition subtype.is_inhabited {a : A} (h : P a) : inhabited {x \ P x} :=
⟨⟨a, h⟩⟩
