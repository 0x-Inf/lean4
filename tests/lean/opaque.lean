import macros
import pair
import subtype
import optional
using subtype
using optional

namespace sum

set_opaque optional false
set_opaque subtype false
set_opaque some false
set_opaque optional::none false
set_opaque rep false
set_opaque subtype::abst false
set_opaque optional_pred false

end
