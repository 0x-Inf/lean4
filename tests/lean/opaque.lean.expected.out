  Set: pp::colors
  Set: pp::unicode
  Imported 'macros'
  Imported 'subtype'
  Imported 'optional'
  Using: subtype
  Using: optional
